File: src/content/docs/topics/cs/coding-problems/tries/211-design-add-and-search-words-data-structure-approach2.go
package main

import "fmt"

// Approach 2: length-bucketed map + per-position char compare

type WordDictionary struct {
	byLen map[int][]string
}

func NewWordDictionary() *WordDictionary {
	return &WordDictionary{byLen: make(map[int][]string)}
}

func (wd *WordDictionary) AddWord(word string) {
	wd.byLen[len(word)] = append(wd.byLen[len(word)], word)
}

func (wd *WordDictionary) Search(word string) bool {
	candidates := wd.byLen[len(word)]
	for _, w := range candidates {
		match := true
		for i := 0; i < len(word); i++ {
			if word[i] != '.' && word[i] != w[i] {
				match = false
				break
			}
		}
		if match {
			return true
		}
	}
	return false
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func runTests() {
	wd := NewWordDictionary()
	wd.AddWord("bad")
	wd.AddWord("dad")
	wd.AddWord("mad")
	assert(wd.Search("pad") == false)
	assert(wd.Search("bad") == true)
	assert(wd.Search(".ad") == true)
	assert(wd.Search("b..") == true)
	assert(wd.Search("...") == true)
	assert(wd.Search("....") == false)
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/tries/211-design-add-and-search-words-data-structure-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

class WordDictionary {
    private byLen: Map<number, string[]> = new Map();

    addWord(word: string): void {
        if (!this.byLen.has(word.length)) this.byLen.set(word.length, []);
        this.byLen.get(word.length)!.push(word);
    }

    search(word: string): boolean {
        const candidates = this.byLen.get(word.length) ?? [];
        for (const w of candidates) {
            let match = true;
            for (let i = 0; i < word.length; i++) {
                if (word[i] !== '.' && word[i] !== w[i]) { match = false; break; }
            }
            if (match) return true;
        }
        return false;
    }
}

const wd = new WordDictionary();
wd.addWord("bad"); wd.addWord("dad"); wd.addWord("mad");
assert(wd.search("pad") === false);
assert(wd.search("bad") === true);
assert(wd.search(".ad") === true);
assert(wd.search("b..") === true);
assert(wd.search("...") === true);
assert(wd.search("....") === false);
console.log("all tests pass");
